community/kernel/src/test/java/org/neo4j/kernel/ShutdownRaceTest.java

```diff
@@ -0,0 +1,150 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.neo4j.kernel;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.neo4j.function.Function;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.kernel.impl.util.function.Optional;
+import org.neo4j.test.Barrier;
+import org.neo4j.test.DatabaseRule;
+import org.neo4j.test.EmbeddedDatabaseRule;
+import org.neo4j.test.ThreadingRule;
+import org.neo4j.tooling.GlobalGraphOperations;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+import static org.neo4j.helpers.collection.Iterables.count;
+import static org.neo4j.kernel.impl.util.function.Optionals.flatMapFutures;
+import static org.neo4j.kernel.impl.util.function.Optionals.none;
+import static org.neo4j.kernel.impl.util.function.Optionals.some;
+
+public class ShutdownRaceTest
+{
+    private static final Function<GraphDatabaseService, Long> COUNT_NODES = new Function<GraphDatabaseService, Long>()
+    {
+        @Override
+        public Long apply( GraphDatabaseService db )
+        {
+            return count( GlobalGraphOperations.at( db ).getAllNodes() );
+        }
+    };
+    public final @Rule DatabaseRule db = new EmbeddedDatabaseRule();
+    public final @Rule ThreadingRule threading = new ThreadingRule();
+    private final Function<Barrier, Optional<Throwable>> createNode = new Function<Barrier, Optional<Throwable>>()
+    {
+        @Override
+        public Optional<Throwable> apply( Barrier barrier )
+        {
+            GraphDatabaseService graphDb = db.getGraphDatabaseService();
+            try ( Transaction tx = graphDb.beginTx() )
+            {
+                barrier.reached();
+                graphDb.createNode();
+
+                tx.success();
+            }
+            catch ( Throwable e )
+            {
+                return some( e );
+            }
+            return none();
+        }
+
+        @Override
+        public String toString()
+        {
+            return "createNode";
+        }
+    }, restart = new Function<Barrier, Optional<Throwable>>()
+    {
+        @Override
+        public Optional<Throwable> apply( Barrier barrier )
+        {
+            try
+            {
+                barrier.reached();
+                db.restartDatabase();
+            }
+            catch ( Throwable e )
+            {
+                return some( e );
+            }
+            return none();
+        }
+
+        @Override
+        public String toString()
+        {
+            return "restart";
+        }
+    };
+
+    @Test
+    public void shouldApplyAllCommittedTransactionsBeforeShuttingDown() throws Exception
```

Comment:
The other test case (`ShutdownAndCommitRaceTest`) triggers the issue directly, whereas this test case needs to run in a debugger with carefully controlled timing to trigger it reliably.

Perhaps we don't need to include this test case (and the changes that go with it).